### ISSUE TITLE
Error: stream.push() after EOF

### ISSUE THREAD

**Issue by florianlenz** [awslabs/amazon-qldb-driver-nodejs#7]
Hi,

I am running into the following error when using this package:

```
(node:97774) UnhandledPromiseRejectionWarning: Error: stream.push() after EOF
    at readableAddChunk (_stream_readable.js:240:30)
    at ResultStream.Readable.push (_stream_readable.js:208:10)
    at ResultStream.<anonymous> (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:141:39)
    at step (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:58:23)
    at Object.next (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:39:53)
    at fulfilled (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:30:58)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:229:7)
(node:97774) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:97774) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:97774) UnhandledPromiseRejectionWarning: Error: stream.push() after EOF
    at readableAddChunk (_stream_readable.js:240:30)
    at ResultStream.Readable.push (_stream_readable.js:208:10)
    at ResultStream.<anonymous> (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:141:39)
    at step (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:58:23)
    at Object.next (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:39:53)
    at fulfilled (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:30:58)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:229:7)
(node:97774) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:97774) UnhandledPromiseRejectionWarning: Error: stream.push() after EOF
    at readableAddChunk (_stream_readable.js:240:30)
    at ResultStream.Readable.push (_stream_readable.js:208:10)
    at ResultStream.<anonymous> (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:141:39)
    at step (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:58:23)
    at Object.next (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:39:53)
    at fulfilled (/Users/florian/projects/qldb/node_modules/amazon-qldb-driver-nodejs/dist/src/ResultStream.js:30:58)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:229:7)
(node:97774) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
[ 'my_table', 'my_tble_two', 'my_table_two' ]
```

the error is causes by this code:

```typescript
const driver = new PooledQldbDriver(
    "my-ledger",
    {
        region: "eu-central-1"
    },
    0,
    500,
    3000,
);

(async () => {

    const s1 = await driver.getSession();
    await s1.executeStatement(`CREATE TABLE my_tble_two`, []);
    console.log(await s1.getTableNames());

})()
    .then(() => {
        console.log("done")
    })
    .catch((error) => {
        console.log(error)
    });
```

I checked the docs again and it doesn't appear to me that i am misusing this library. I also wonder about the `UnhandledPromiseRejectionWarning` since I am handing the rejected promise (and actually, you can see that "done" is printed, so my `.then` handler is called).

Let me know if I can do anything to help solve / debug this issue.

**Comment by saumehta9** [comment]
Hi Florian, 

Your code seems to be absolutely okay. We would need some more information from your side. I tried reproducing this on my end, but it worked absolutely fine with no warnings or errors. There is probably a difference in the installations.

1. Can you share the detailed steps you followed to install (things like aws sdk, qldb driver, ion) and get things up and running.
2. How are you importing the dependencies ?
3. Do you have any global installations of dependencies like qldb driver, aws-sdk, ion.
4. Did you notice any errors/warning regarding mismatching peer dependencies.
5. Also please share your dependency config - package.json and package-lock.json
6. Are you doing any npm links to global dependencies such as aws sdk, ion ?

Also you mentioned that "done" is being logged, but i could not really see that in the logs. Though looking at your code and logs it makes perfect sense, but still can you confirm it is being logged ?

**Comment by florianlenz** [comment]
Hi,

> 1. Can you share the detailed steps you followed to install (things like aws sdk, qldb driver, ion) and get things up and running.
I attached a test project. There's only one script `run.js` that you can run.

> 2. How are you importing the dependencies ?
In the test project I am using `const {PooledQldbDriver} = require("amazon-qldb-driver-nodejs");`

> 3. Do you have any global installations of dependencies like qldb driver, aws-sdk, ion.
No

> 4. Did you notice any errors/warning regarding mismatching peer dependencies.
Yes, but I installed the missing dependencies, so it should be fine.

> 5. Also please share your dependency config - package.json and package-lock.json
Is included in the test project

> 6. Are you doing any npm links to global dependencies such as aws sdk, ion ?
Not that I am aware of.

> Also you mentioned that "done" is being logged, but i could not really see that in the logs. Though looking at your code and logs it makes perfect sense, but still can you confirm it is being logged ?
Sorry, it's part of the output but I didn't add it by mistake. You will see it when you run the script from the test project.

Let me know if you need anything.

[qldb-example.zip](https://github.com/awslabs/amazon-qldb-driver-nodejs/files/3928492/qldb-example.zip)

**Comment by saumehta9** [comment]
Hi @florianlenz,

Thanks for sharing the code. 
We tried running it. It runs absolutely fine.

Logs:

```
➜  qldb-example node run.js
[
  'another_table',
  'my_tble_1',
  'another_table1',
  'another_table3',
  'my_tble',
  'my_tble_two'
]
done
```
We are unable to reproduce the issue.

**Comment by kylepbit** [comment]
Hi @florianlenz, 

Is this issue consistently reproducible, or is it intermittent?

**Comment by florianlenz** [comment]
@kylepbit this issue is consistent between multiple projects. I just downloaded the test project I created again and I am still running into the same error. So at least on my machine it's consistently reproducible. 
I am using OSX Catalina with node v8.16.2.

@saumehta9 how did you install the dependencies and which OS are you using? Are you using `npm ci` to install the dependencies?

**Comment by saumehta9** [comment]
@florianlenz 

OS - MacOS Mojave v10.14.6
Node version - v13.0.1

I installed it by running npm install on your project.
Here is what my dependency tree looks like:

```
➜  qldb-example npm list 
qldb-example@1.0.0 /Users/saumehta/Downloads/qldb-example
├─┬ amazon-qldb-driver-nodejs@0.1.0-preview.2
│ ├── @types/node@12.0.2
│ ├─┬ ion-hash-js@1.0.3
│ │ └─┬ ion-js@3.1.2
│ │   └── jsbi@3.1.1
│ ├── semaphore-async-await@1.5.1
│ └─┬ stream@0.0.2
│   └── emitter-component@1.1.1
└─┬ aws-sdk@2.585.0
  ├─┬ buffer@4.9.1
  │ ├── base64-js@1.3.1
  │ ├── ieee754@1.1.13 deduped
  │ └── isarray@1.0.0
  ├── events@1.1.1
  ├── ieee754@1.1.13
  ├── jmespath@0.15.0
  ├── querystring@0.2.0
  ├── sax@1.2.1
  ├─┬ url@0.10.3
  │ ├── punycode@1.3.2
  │ └── querystring@0.2.0 deduped
  ├── uuid@3.3.2
  └─┬ xml2js@0.4.19
    ├── sax@1.2.1 deduped
    └── xmlbuilder@9.0.7
```

**Comment by byronlin-bq** [comment]
Hi @florianlenz,

I was able to reproduce your issue using your Node version 8.16.2.

We are currently investigating this issue.

In the meantime, a workaround would be to upgrade your Node version to v10. It seems readable-stream v3 (Node v10) resolved this issue: https://www.nearform.com/blog/welcome-readable-stream-3/#asyncread

**Comment by florianlenz** [comment]
Great!

**Comment by florianlenz** [comment]
Any news on the resolution of this issue? We sadly can't upgrade to v10 in production and are stuck with version 8.

**Comment by aurghob33** [comment]
Hi @florianlenz. 
Thank you for reaching out. We’re working on a fix. Please stay tuned here for further updates.

**Comment by aurghob33** [comment]
Hi @florianlenz, 
The change is published both to Github and npm with version 0.1.2-preview.1. Do let us know if it works for you.

**Comment by aurghob33** [comment]
Closing this issue since it has been open for a month with no feedback and the change is pushed through to npm as well. If there are any further questions, feedback or concerns, please feel free to re-open the issue!
